matminer/featurizers/composition/roost.py

```diff
@@ -0,0 +1,569 @@
+        #         )
+        #         if self.model.task == "regression":
+        #             val_score = v_metrics["MAE"]
+        #             print(f"Validation Baseline: MAE {val_score:.3f}\n")
+        #         elif self.model.task == "classification":
+        #             val_score = v_metrics["Acc"]
+        #             print(f"Validation Baseline: Acc {val_score:.3f}\n")
+        #         self.model.best_val_score = val_score
+
+        if checkpoint:
+            if not os.path.isdir("models/"):
+                os.makedirs("models/")
+
+            if not os.path.isdir(f"models/{self.model_name}/"):
+                os.makedirs(f"models/{self.model_name}/")
+
+        if log:
+            if not os.path.isdir("runs/"):
+                os.makedirs("runs/")
+
+            writer = SummaryWriter(
+                # NOTE comprhys: I find it useful to automatically label by
+                # time and date but haven't added this here as I imagine that
+                # people won't do as much experimentation with hypers and they
+                # can therefore use model_name to differentiate
+                log_dir=f"runs/{self.model_name}-r{self.run_id}"
+            )
+        else:
+            writer = None
+
+        self.model.fit(
+            train_generator=train_generator,
+            val_generator=None,
+            # val_generator=val_generator,
+            optimizer=self.optimizer,
+            scheduler=self.scheduler,
+            epochs=self.epochs,
+            criterion=self.criterion,
+            normalizer=self.normalizer,
+            model_name=self.model_name,
+            run_id=self.run_id,
+            checkpoint=checkpoint,
+            writer=writer,
+            verbose=verbose,
+        )
+
+        self.fitted = True
+
+        # NOTE move model to cpu as featurisation is faster on the cpu unless
+        # data is batched. Currently featurise_many doesn't use batches and
+        # instead applies the featuriser to each row in turn.
+        self.model.device = "cpu"
+        self.model.to("cpu")
+
+    def featurize(self, *comp):
+
+        if not self.fitted:
+            raise ValueError(f"Please fit the {self} before use")
+        test_set = CompositionData(comp, task=self.task)
+
+        test_generator = DataLoader(test_set, **self.data_params)
+
+        # Ensure model is in evaluation mode
+        self.model.eval()
+
+        with torch.no_grad():
+            for input_, _, _, _ in test_generator:
+                # move tensors to device
+                input_ = (tensor.to(self.model.device) for tensor in input_)
+                # compute intermediate output
+                output = self.model.material_nn(*input_).numpy().ravel()
+                # output = self.model.material_nn(*input_).data.cpu().numpy().ravel()
+
+        return output
+
+    def feature_labels(self):
+        return [f"Roost_feature_{x}" for x in range(self.elem_fea_len)]
```

Comment:
Does this featurizer return a learned embedding of the material, or a prediction on the property on interest?

Reply:
So I broke the model into two modules, the descriptor module which is the core of the method and then an output nn. There's nothing special about the output nn so I thought it would be more in keeping with `matminer` to return the features from the descriptor module.

For transparency from some recent chemoinformatics work I have reason to believe that trained `roost` features + random forests should perform very well on small datasets in particular. So I was making choices based on how to make this use case as easy as possible when setting up the `RoostFeaturizer`